x-pack/metricbeat/module/aws/s3_daily_storage/s3_daily_storage.go

```diff
@@ -0,0 +1,220 @@
+type MetricSet struct {
+	*aws.MetricSet
+	logger *logp.Logger
+}
+
+// New creates a new instance of the MetricSet. New is responsible for unpacking
+// any MetricSet specific configuration options if there are any.
+func New(base mb.BaseMetricSet) (mb.MetricSet, error) {
+	cfgwarn.Beta("The aws s3_daily_storage metricset is beta.")
+	s3Logger := logp.NewLogger(aws.ModuleName)
+
+	moduleConfig := aws.Config{}
+	if err := base.Module().UnpackConfig(&moduleConfig); err != nil {
+		return nil, err
+	}
+
+	if moduleConfig.Period == "" {
+		err := errors.New("period is not set in AWS module config")
+		s3Logger.Error(err)
+	}
+
+	metricSet, err := aws.NewMetricSet(base)
+	if err != nil {
+		return nil, errors.Wrap(err, "error creating aws metricset")
+	}
+
+	// Check if period is set to be multiple of 86400s
+	remainder := metricSet.PeriodInSec % 86400
+	if remainder != 0 {
+		err := errors.New("period needs to be set to 86400s (or a multiple of 86400s). " +
+			"To avoid data missing or extra costs, please make sure period is set correctly " +
+			"in config.yml")
+		s3Logger.Info(err)
+	}
+
+	return &MetricSet{
+		MetricSet: metricSet,
+		logger:    s3Logger,
+	}, nil
+}
+
+// Fetch methods implements the data gathering and data conversion to the right
+// format. It publishes the event which is then forwarded to the output. In case
+// of an error set the Error field of mb.Event or simply call report.Error().
+func (m *MetricSet) Fetch(report mb.ReporterV2) {
+	namespace := "AWS/S3"
+	// Get startTime and endTime
+	startTime, endTime, err := aws.GetStartTimeEndTime(m.DurationString)
+	if err != nil {
+		logp.Error(errors.Wrap(err, "Error ParseDuration"))
+		m.logger.Error(err.Error())
+		report.Error(err)
+		return
+	}
+
+	// GetMetricData for AWS S3 from Cloudwatch
+	for _, regionName := range m.MetricSet.RegionsList {
+		m.MetricSet.AwsConfig.Region = regionName
+		svcCloudwatch := cloudwatch.New(*m.MetricSet.AwsConfig)
+		listMetricsOutputs, err := aws.GetListMetricsOutput(namespace, regionName, svcCloudwatch)
+		if err != nil {
+			m.logger.Error(err.Error())
+			report.Error(err)
+			continue
+		}
+
+		if listMetricsOutputs == nil || len(listMetricsOutputs) == 0 {
+			continue
+		}
+
+		metricDataQueries := constructMetricQueries(listMetricsOutputs, m.PeriodInSec)
+		// Use metricDataQueries to make GetMetricData API calls
+		metricDataOutputs, err := aws.GetMetricDataResults(metricDataQueries, svcCloudwatch, startTime, endTime)
+		if err != nil {
+			err = errors.Wrap(err, "GetMetricDataResults failed, skipping region "+regionName)
+			m.logger.Error(err.Error())
+			report.Error(err)
+			continue
+		}
+
+		// Create Cloudwatch Events for s3_daily_storage
+		bucketNames := getBucketNames(listMetricsOutputs)
+		for _, bucketName := range bucketNames {
+			event, err := createCloudWatchEvents(metricDataOutputs, regionName, bucketName)
+			if err != nil {
+				m.logger.Error(err.Error())
+				event.Error = err
+				report.Event(event)
```

Comment:
Shouldn't this be `report.Error(err)`?

Reply:
No 😄, what I want to do here is to report the event with whatever data it has AND an error. So I don't lose the whole event when an error occurred.

Reply:
Ok :smile: I was just double checking :)